Refactors validation of config and passes arguments to onwarn correctly, fixes #1762

ankeetmaini opened the pull request:
Currently, the options were validated for deprecations separately for CLI usage with config and Node API. I've tried to make them uniform by using `deprecateOptions` file. This is a little involved because, in CLI, config includes both input and output options, whereas in Node API we pass them separately.

To accomplish this I've split the `deprecateOptions` internally into two functions one for input options and another for output so that we could use this file irrespective of the client invoked(CLI, Node API).

I've also added a test that correct arguments are passed to `onwarn` function when using rollup.rollup() but couldn't figure out how to add it for CLI.

I've also deleted specific checks from `src/rollup/index.js` as they were covered in `deprecateOptions`.

@lukastaegert please let me know if this is the right way to approach this.

ankeetmaini commented:
Oops, build failed, I'll fix it.

adrianheine commented on bin/src/run/batchWarnings.js:
```diff
@@ -212,13 +212,16 @@ const deferredHandlers = {
-			nestedByPlugin.forEach(({ key: plugin, items }) => {
+			nestedByPlugin.forEach(nested => {
+				const items = nested.items;
+				const plugin = nested.key;
```
Why did you replace this destructuring?

ankeetmaini replied:
I did this because it was failing build on Node 4, `rollup.config.js` uses 
```js
buble({include: ['src/**', 'node_modules/acorn/**']})
```
so the same code retains destructuring in rollup.js whereas it removes in the final built file of /bin/rollup

[Build failure link](https://travis-ci.org/rollup/rollup/builds/310089193?utm_source=github_status&utm_medium=notification)

lukastaegert replied:
Maybe we should change this to include the `bin` path as well. I don't see why not.

ankeetmaini replied:
As far as I can understand including the path in `buble` keeps the source code as is and doesn't transpile down. So if we were to include `bin` path it'll also break the build for Node 4 for destructured code as Node 4 doesn't support it.

The reason `bin/rollup` bundle works is because it gets the code correctly transpiled down and all the destructuring is removed.

lukastaegert replied:
Ah, I see, I think I understand. So the problem is that now for the first time, files are imported from the `bin` folder by modules in the `src` folder.

I think the solution is simple: Common files used by both the `bin` folder and the `src` folder should got into the `src` folder. More specifically, I think that `batchWarnings.js` and `deprecateOptions` should both go into `src/utils`. This is following a pattern established for files like `relativeId.js`, `promise.js` or `ensureArray.js`. By keeping exports one way we make sure that we can always freely refactor inside the `bin` folder without possibly breaking functionality elsewhere. Then I guess you should be able to put the destructuring back in.

BTW I really like what you are doing with this PR, this duplicate deprecation was something that REALLY bugged me.

ankeetmaini commented:
@lukastaegert this is still a work in progress. I think I might've found a better way, as of this PR I'm using deprecateOptions directly whereas I've seen in CLI and watch mode we use `mergeOptions` to merge and calculate deprecations. This way we'll be sure that all the configs are parsed in the same manner.

adrianheine commented on bin/src/run/index.js:
```diff
@@ -1,8 +1,8 @@
-import mergeOptions from './mergeOptions.js';
-import batchWarnings from './batchWarnings.js';
+import mergeOptions from '../../../src/utils/mergeOptions';
+import batchWarnings from '../../../src/utils/batchWarnings';
```
Please keep the file extensions.

adrianheine commented on bin/src/run/loadConfigFile.js:
```diff
@@ -1,7 +1,7 @@
 import chalk from 'chalk';
 import * as rollup from 'rollup';
-import batchWarnings from './batchWarnings.js';
+import batchWarnings from '../../../src/utils/batchWarnings';
```
Please keep the file extension.

adrianheine commented on bin/src/run/watch.js:
```diff
@@ -4,8 +4,8 @@ import chalk from 'chalk';
-import mergeOptions from './mergeOptions.js';
-import batchWarnings from './batchWarnings.js';
+import mergeOptions from '../../../src/utils/mergeOptions';
+import batchWarnings from '../../../src/utils/batchWarnings';
```
Please keep the file extensions.

adrianheine commented on src/rollup/index.js:
```diff
@@ -3,53 +3,14 @@ import { basename } from '../utils/path.js';
-import deprecateOptions from '../../bin/src/run/deprecateOptions';
-import batchWarnings from '../../bin/src/run/batchWarnings';
+import mergeOptions from '../utils/mergeOptions';
+import batchWarnings from '../utils/batchWarnings';
```
Please add the file extensions.

adrianheine commented on src/utils/batchWarnings.js:
```diff
@@ -1,6 +1,6 @@
-import { stderr } from '../logging.js';
-import relativeId from '../../../src/utils/relativeId.js';
+import { stderr } from '../../bin/src/logging';
+import relativeId from './relativeId';
```
Please keep the file extensions.

adrianheine replied:
Also, this still imports a `bin` module from `src`, right?

ankeetmaini commented:
Following is the gist of things I've done:
- [x] moved `deprecateOptions` from `bin/src/run` → `src/utils`
- [x] moved `batchWarnings.js` from `bin/src/run` → `src/utils`
- [x] moved `mergeOptions.js` from `bin/src/run` → `src/utils`
- [x] removed onwarn decalarations to `mergeOptions.js`
- [x] removed `allowedKeys` whitelisting in `src/rollup/index.js` as everything can now be done in `mergeOptions.js` as a single source of truth
- [x] similarly removed whitelisting of input/output options from `src/watch/index.js`
- [ ] **make tests pass, there 4 tests failing**

This is the general approach I've to make validation of config unified accross CLI, watch and rollup.rollup. If this change is too large please let me know if I should break it, or if you're happy with this approach.

lukastaegert commented:
As for myself, I am always in favour of larger refactorings that clean up stuff 😜. Probably won't have time to take a look before next week, though. Input from others is always welcome, e.g. @adrianheine 😉

ankeetmaini commented:
This is ready to be reviewed now. All tests passing :D

adrianheine commented on src/utils/batchWarnings.js:
```diff
@@ -1,6 +1,6 @@
 import chalk from 'chalk';
-import { stderr } from '../logging.js';
-import relativeId from '../../../src/utils/relativeId.js';
+import { stderr } from '../../bin/src/logging.js';
```
This statement in `src/` imports from `bin/`. Maybe it's possible to pass in `stderr`?

ankeetmaini commented:
I'll do some more finding and create a test for the unified new config

ankeetmaini commented:
@lukastaegert  I've made the following changes as you suggested
- [x] tests for testing deprecation logics for CLI and Node
- [x] redefined the `stderr` function in src folder
- [x] debugged the `undefined` printouts in console, they were happening as we weren't passing the message attribute to onwarn, and when a custom onwarn isn't present we just print warning.message. Earlier there weren't any deprecations so it was running just fine.

lukastaegert commented on src/watch/index.js:
```diff
@@ -142,7 +110,7 @@ class Task {
 
 		this.filter = createFilter(watchOptions.include, watchOptions.exclude);
-		this.deprecations = watchOptions._deprecations;
+		this.deprecations = [...deprecations, watchOptions._deprecations];
```
It seems to me that this should actually read `[...deprecations, ...watchOptions._deprecations]`?
If so, maybe filtering for truthy deprecations is not necessary after all?

lukastaegert commented:
Hi @ankeetmaini,
I am REALLY sorry but even though it is marked as "merged", I have to ask you to re-open this PR (or probably a new PR from this branch). After I merged your branch into the release branch and ran the tests, I stumbled upon a critical problem (and a minor issue) in this branch that prevents me from merging it to master.

You will see the warning yourself when you run the tests on your branch:
```
(!) Missing shims for Node.js built-ins
Creating a browser bundle that depends on 'os'. You might need to include https://www.npmjs.com/package/rollup-plugin-node-builtins
(!) Unresolved dependencies
https://github.com/rollup/rollup/wiki/Troubleshooting#treating-module-as-external-dependency
os (imported by node_modules/chalk/node_modules/supports-color/index.js, commonjs-external:os)
(!) Missing global variable name
Use options.globals to specify browser global variable names corresponding to external modules
os (guessing 'os')
```

The long and short of this warning is this: Due to the fact that `batchWarnings.js` is now directly used by the node API, the node API depends on `chalk` which is not browser compatible (at least not out of the box). Browser compatibility, however, is crucial for e.g. running rollup in the REPL.

I have dug a little further and I think `batchWarnings.js` was always meant as a CLI library; most of its purpose is to properly format errors in a CLI environment. Therefore, it should go back into the bin folder. Instead, `mergeOptions.js` needs to lose its direct dependency on `batchWarnings.js`. Instead, warnings should be treated like all other warnings in the node API: Be passed to the `onwarn` handler provided by the caller (which in case of the CLI uses `batchWarnings.js` and thus has the nice formatting) or be logged to console.error without formatting. I think a little more digging needs to be done to figure out what is best.

Another issue that is not as critical (and that I overlooked as well) is the fact that we no longer have any "typo-protection" any more. Previously when the user mistyped an option, there would be a warning about using an unknown option. Now there is just silence and the user will wonder why the option does not have the intended effect.
I know the previous solution was not ideal in that it contained a white-list that needed to be hand-curated. However, IMO this is still better than the current solution of being quiet about it. If we want to improve things here, we could curate two different lists of options, one for input and one for output.

I am sorry I did not notice these things earlier, I was a little short on time the previous weeks and should have looked more carefully.

ankeetmaini commented:
@lukastaegert no worries! I will fix the issue of dependency on batchWarnings.js, also will see how can we solve the issue of typo-warnings. Might take a little time :)

I'll reopen this PR maybe from master, if you plan on releasing 0.53  soon!

lukastaegert commented:
Yes, please reopen it from master! I only create and use release branches directly before a release to be able to do some final checks before merging to master.

Thanks for having another go. Take as much time as you need, you can always for feedback!
